### PR TITLE
Update field partials for correct use of options variables

### LINKED ISSUE #631: Rename option-related variables to reflect vanilla Rails conventions
After some discussion in #587, it has become clear that we should use our options like this:
1. `options`: Options for buttons/select fields/etc. For example, `["one", "two", "three"]`.
2. `html_options`: The same as the Rails helpers in that we set the class attribute, etc.
3. `other_options`: Bullet Train specific options. For example `help`, `hide_label`, and `require`.

Read [this comment](https://github.com/bullet-train-co/bullet_train-core/pull/587/files#r1368571161) for more details.

@jagthedrummer suggested that we use either `bt_options` or `fp_options` instead of `other_options`. I think this is a good idea, and that it will be clearer for everyone to understand.

## #587
I think we should figure this out before #587, because I foresee this being a pretty big PR and it might be better organize things mentally before merging the `disable` logic from there.

### PR DESCRIPTION
**Pull request by gazayas**
Closes #631.

The current state of this PR doesn't change the use of `options` and `html_options`, but it does remove them from render calls where they aren't necessary. I've added comments to those invocations to make the PR easier to review, but we can delete them later if we decide we don't need them.

- [x] Add docs for `options` vs `html_options`
- [x] Write comments in the code for clarity
- [x] Remove `options` or `html_options` where necessary in `bullet_train-themes` partials
- [x] Remove `options` or `html_options` where necessary in `bullet_train-themes-tailwind-css` partials
- [x] Update Scaffolding::Transformer to scaffold options correctly if necessary
- [ ] Change all instances of `other_options` to `bt_options`
- [x] Add any deprecation messages where necessary
- [x] Check showcase partials if they need to be updated
- [ ] Add a page to the documentation for updating to the next version

There are a lot of tasks here, so I'm okay with breaking down this one into pieces if we need to.

## _field.html.erb
These two files are seemingly the same, so I wonder if we can consolidate this all into one file.
1. [bullet_train-themes-light/app/views/themes/light/fields/_field.html.erb](https://github.com/bullet-train-co/bullet_train-core/blob/main/bullet_train-themes-light/app/views/themes/light/fields/_field.html.erb)
2. [bullet_train-themes-tailwind_css/app/views/themes/tailwind_css/fields/_field.html.erb](https://github.com/bullet-train-co/bullet_train-core/blob/main/bullet_train-themes-tailwind_css/app/views/themes/tailwind_css/fields/_field.html.erb)

EDIT:
The tailwind_css field partial was a duplicate, so I deleted it here https://github.com/bullet-train-co/bullet_train-core/pull/633/commits/f3fdcee324a0aac86725000589f77130f7dd9439

### CODE REVIEW

**Comment by gazayas** [comment]
I went ahead and deleted `bullet_train-themes-tailwind_css/app/views/themes/tailwind_css/fields/_field.html.erb`, it looks like we didn't actually need it.

**Comment by gazayas** [comment]
We'll have to look over this block to make sure we're scaffolding `multiple: true` correctly.

https://github.com/bullet-train-co/bullet_train-core/blob/cdccbdd270c62592458447e9ee0af5550fdb781c/bullet_train-super_scaffolding/lib/scaffolding/transformer.rb#L739-L747

**Comment by gazayas** [comment]
Still a handful of partials to work through, but I'm hoping to get this one over the finish line sometime next week. I'm foreseeing the following partials being a little difficult to adjust:

1. `_cloudinary_image.html.erb`
2. `_date_and_time_field.html.erb`
3. `_date_field.html.erb`
4. `_option.html.erb`
5. `_options.html.erb`

**Comment by gazayas** [comment]
Managed to get the cloudinary image partial and color picker partial out of the way today, only a few more partials left.

One thing to note is that color picker options were being housed like this:
`options[:color_picker_options]`

Other places, like the button partial, simply put the array of values into `options` directly (without a key). I've decided to just use an array named `color_picker_field_options`. This is the same as what I've done with `button_field_options`. I hope to unify all of these with the `*_field_options` naming convention.

**Comment by gazayas** [comment]
@jagthedrummer This PR is done for the most part, so I marked it for review as I think over how to add documentation for upgrading to the next version. I know you did a lot of upgrading docs before, so any insight would be helpful.

I also have left the `other_options` to `bt_options` task unchecked for now because there's already a lot going on here. Do you think it should be handled in this PR as well, or should we separate it?

Besides the docs, I'll look over things one more time to see what I can clean up.

**Comment by jagthedrummer** [comment]
I've been pretty tied up this week and haven't had a chance to look at this. I'll give it a good review next week. Thanks for taking this on, @gazayas!

**Comment by gazayas** [comment]
@jagthedrummer No problem at all! Just as a point of reference, here is the main method that calls the Rails form helpers, which is what I think we should base `options` and `html_options` off of.

https://github.com/bullet-train-co/bullet_train-core/blob/6d2ca8a7054d7e4e899ffefadd2238ffacef59ad/bullet_train-themes-light/app/views/themes/light/fields/_field.html.erb#L69-L81

File: bullet_train-super_scaffolding/lib/scaffolding/transformer.rb [hunk 1]
```diff
@@ -754,7 +742,7 @@ def valid_#{attribute.collection_name}
 
         # https://stackoverflow.com/questions/21582464/is-there-a-ruby-hashto-s-equivalent-for-the-new-hash-syntax
         if field_options.any? || options.any?
-          field_options_key = if ["buttons", "super_select", "options"].include?(attribute.type)
```
Comment (gazayas) [inline bullet_train-super_scaffolding/lib/scaffolding/transformer.rb:757]: Super Select is the only partial that uses `html_options` under the hood, whereas buttons and options simply use `options` (all the BT-side invocations are taken from `main`).

Super Select:
- [select API](https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-select) and [BT-side invocation](https://github.com/bullet-train-co/bullet_train-core/blob/cdccbdd270c62592458447e9ee0af5550fdb781c/bullet_train-themes/app/views/themes/base/fields/_super_select.html.erb#L46)

Buttons:
- [check_box API](https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-check_box) and [BT-side invocation](https://github.com/bullet-train-co/bullet_train-core/blob/cdccbdd270c62592458447e9ee0af5550fdb781c/bullet_train-themes-tailwind_css/app/views/themes/tailwind_css/fields/_buttons.html.erb#L18)
- [radio_button API](https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-radio_button) and [BT-side invocation](https://github.com/bullet-train-co/bullet_train-core/blob/cdccbdd270c62592458447e9ee0af5550fdb781c/bullet_train-themes-tailwind_css/app/views/themes/tailwind_css/fields/_buttons.html.erb#L20)

Options:
- [check_box](https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-check_box) and [BT-side invocation](https://github.com/bullet-train-co/bullet_train-core/blob/cdccbdd270c62592458447e9ee0af5550fdb781c/bullet_train-themes-tailwind_css/app/views/themes/tailwind_css/fields/_option.html.erb#L23)
- [check_box_tag](https://api.rubyonrails.org/classes/ActionView/Helpers/FormTagHelper.html#method-i-check_box_tag) and [BT-side invocation](https://github.com/bullet-train-co/bullet_train-core/blob/cdccbdd270c62592458447e9ee0af5550fdb781c/bullet_train-themes-tailwind_css/app/views/themes/tailwind_css/fields/_option.html.erb#L29)
- [radio_button](https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-radio_button) and [BT-side invocation](https://github.com/bullet-train-co/bullet_train-core/blob/cdccbdd270c62592458447e9ee0af5550fdb781c/bullet_train-themes-tailwind_css/app/views/themes/tailwind_css/fields/_option.html.erb#L41)

File: bullet_train-themes-light/app/views/themes/light/fields/_field.html.erb [hunk 2]
```diff
@@ -30,14 +50,14 @@ end
...
 
+# TODO: Partials which invoke form helpers inline aren't using
+# these classes, although we have the `has_errors` conditional with some style below.
 options[:class] = "#{options[:class]} block w-full rounded-md shadow-sm font-light text-sm dark:bg-slate-800 dark:text-slate-300"
-
 options[:class] += if has_errors
   " pr-10 border-red-500 text-red-900 placeholder-red-500 outline-none focus:ring-red-500 focus:border-red-500 dark:text-slate-300"
 else
   " focus:ring-primary-500 focus:border-primary-500 border-slate-300 dark:border-slate-900"
 end
```
Comment (gazayas) [inline bullet_train-themes-light/app/views/themes/light/fields/_field.html.erb:60]: Example of the proper classes not being added to a `date_and_time_field` named `foo` although we're passing `options` to `render 'shared/fields/field'`. I'm thinking of making a helper to compensate for this.
スクリーンショット 2023-11-11 17 06 05

**Comment by gazayas** [comment]
I've only edited comments for these last three commits, it looks like the Open API tests are acting up again.

**Comment by gazayas** [comment]
One more note before I call it a day. I decided to pass all `options` to `render 'shared/fields/field'` for now, but I think we can remove them in the future. It's needed right now for two reasons:
1. For date-related partials to work (it only needs `options[:id]` though, not the whole hash).
2. To add specific classes to the form element, but this is only working for partials that invoke [`form.send(helper, method, options)`](https://github.com/bullet-train-co/bullet_train-core/blob/cdccbdd270c62592458447e9ee0af5550fdb781c/bullet_train-themes-light/app/views/themes/light/fields/_field.html.erb#L62) (details [here](https://github.com/bullet-train-co/bullet_train-core/pull/633#discussion_r1390173709)).

I think we can find a way around these two issues and eventually remove `options: options` from `render 'shared/fields/field'` altogether. However, we don't *have* to worry about it for the scope of this PR.

**Comment by jagthedrummer** [comment]
I just merged a PR to lock `jbuilder-schema` to a known good version, so I'm going to rebase this branch and force push. https://github.com/bullet-train-co/bullet_train-core/pull/659

File: bullet_train-themes/app/views/themes/base/fields/_ckeditor.html.erb [hunk 3]
```diff
@@ -9,9 +9,9 @@ stimulus_controller = 'fields--ckeditor'
 
 form ||= current_fields_form
 options ||= {}
-options = options.merge({ data: {"#{stimulus_controller}-target": 'field' }})
 other_options ||= {}
 
+options = options.merge({ data: {"#{stimulus_controller}-target": 'field' }})
```
Comment (jagthedrummer) [inline bullet_train-themes/app/views/themes/base/fields/_ckeditor.html.erb:14]: This file seems like a remnant of a feature that we no longer support. Maybe we should remove it in the near future?

**Comment by jagthedrummer** [comment]
@gazayas I just pulled this down and tried it out and I think it's looking good in terms of making all of the `options` stuff more consistent.

My one concern (which is a big one) is that the deprecation warnings don't point to a useful place.

First I generated a `Project` model with a bunch of fields to test against.

```
bin/super-scaffold crud Project Team \
  name:text_field \
  logo:file_field \
  address:address_field \
  a_buttons:buttons \
  some_buttons:buttons{multiple} \
  a_color:color_picker \
  a_date_with_time:date_and_time_field \
  a_date:date_field an_option:options \
  some_options:options{multiple} \
  trix_editor_area:trix_editor \
  a_password:password_field \
  phone_number:phone_field \
  a_super_select:super_select \
  some_super_selects:super_select{multiple}
```

Then I ran `bin/hack --link ../bullet_train-core` to link my starter repo to this branch in my `core` repo.

And then I visited the form for editing a project, and I see two deprecation messages that look like this:

```
DEPRECATION WARNING: The `multiple` attribute will be removed in a later version.
  Please pass `options: {multiple: true}` to `render` instead.
  (called from render at ../bullet_train-core/bullet_train-themes/app/helpers/theme_helper.rb:19)
```

The warnings point to `theme_helper.rb` instead of to my project form. And they're both identical even though they're coming from different places. (One from `buttons` and one from `options`.)

Through trial and error I figured out that it's the `options` partial that is raising one of the deprecation messages so I inspected the call stack at that point and the relevant bits looks like this:

```
# bunch of uninteresting lines
../bullet_train-core/bullet_train-themes/app/helpers/theme_helper.rb:19:in `render'
./app/views/account/projects/_form.html.erb:17:in `block (3 levels) in _app_views_account_projects__form_html_erb___3804842824792311178_113380'
../bullet_train-core/bullet_train-fields/app/helpers/fields_helper.rb:13:in `with_field_settings'
# bunch more uninteresting lines
```

With all the extra paths removed, and commentary added:

```
theme_helper.rb:19 <-- We're pointing here
projects/_form.html.erb:17 <-- We should point here
fields_helper.rb:13
```

I figured that the backtrack cleaner must be seeing `theme_helper.rb` as being a "local" file and reporting that, so I tried setting my `Gemfile` to use the released version of `bullet_train-themes` so that `theme_helper.rb` would no longer be considered local.

After doing that the deprecation message looks like this:

```
DEPRECATION WARNING: The `multiple` attribute will be removed in a later version.
  Please pass `options: {multiple: true}` to `render` instead.
  (called from with_field_settings at ../bullet_train-core/bullet_train-fields/app/helpers/fields_helper.rb:13)
```

So it completely skipped over the line in the call stack for `projects/_form.html.erb`. I'm guessing this has something to do with the Rails back trace cleaner not reporting on `erb` files, but I'm not entirely sure of what to do about it.

```
theme_helper.rb:19
projects/_form.html.erb:17 <-- We should point here
fields_helper.rb:13 <-- But now we're pointing here
```

I tried also pointing my `Gemfile` to the released version of `bullet_train-fields` but that just made it so that the deprecation pointed to some `ActionView` internals at `/Users/jgreen/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/actionview-7.0.8/lib/action_view/base.rb:244`.

I tried commenting out portions of `config/initializers/backtrace_silencers.rb` but that didn't seem to have any effect on anything related to this.

It would be nice if we could somehow tell the deprecation plumbing that we _do_ want to include `.erb` files in the deprecation trace, but I'm not sure how to do that. If there's not a way to do that we may need to inspect the call stack ourselves to find out where it's being called from.

In the case of `multiple` being passed to the `options` partial I think an ideal message would look something like this:

```
DEPRECATION WARNING: Passing `multiple` directly to the `shared/field/options` partial is deprecated.
  Please pass `options: {multiple: true}` to `render` instead.
  (called from app/views/account/projects/_form.html.erb:17)
```

**Comment by jagthedrummer** [comment]
I tried to see if the deprecation weirdness is due to something we're doing in BT and I don't think it is. In a brand new vanilla rails app I can't get a deprecation triggered from a partial to report the right location. I've opened an issue on `rails/rails`: https://github.com/rails/rails/issues/50047

**Comment by gazayas** [comment]
> I tried to see if the deprecation weirdness is due to something we're doing in BT and I don't think it is. In a brand new vanilla rails app I can't get a deprecation triggered from a partial to report the right location. I've opened an issue on `rails/rails`: [rails/rails#50047](https://github.com/rails/rails/issues/50047)

I did notice it was referring to the theme helper and not the partial, but it slipped my mind to report on it. Thanks for digging into it!

**Comment by gazayas** [comment]
Fixed in https://github.com/rails/rails/pull/50053

File: bullet_train-themes/app/views/themes/base/fields/_ckeditor.html.erb (hunk 3 above)
Comment (gazayas) [reply bullet_train-themes/app/views/themes/base/fields/_ckeditor.html.erb:15]: Just added a TODO in https://github.com/bullet-train-co/bullet_train-core/pull/633/commits/660bc41d3b40b23074b9b0de11ebcd2a3c439e73 to remind as later (I don't have a proper solution off the top of my head).

**Comment by gazayas** [comment]
I added some more documentation explaining `options` and `other_options` in more detail, along with the general `other_options` that are available for all of the field partials. I think this information will be useful for developers so they know exactly where their information is going (a.k.a what form helper is being used), and we can use it as a guideline in the future in case we're not sure what options or values we want to pass to partials on the Bullet Train side.

**Comment by jagthedrummer** [comment]
@gazayas it looks like the rails fix that we were waiting on was released a couple of days ago in `7.1.3` and the starter repo has already been updated (thanks, depfu!). So I think we can pick this one back up. Would you mind looking at the merge conflict when you have some time?

**Comment by gazayas** [comment]
@jagthedrummer It's been a hot minute so there are going to be a lot of commits from the merge here, but the only conflict was from the update in #742 so it shouldn't be too bad. Also, super scaffolding system tests passed locally for me.

**Comment by gazayas** [comment]
Ah, it just consolidated everything into the merge: https://github.com/bullet-train-co/bullet_train-core/pull/633/commits/54518e4ab1c585ca6ff8f8cf5cb632167048c554

Alright, good to go!

**Comment by gazayas** [comment]
Linter...

**Comment by gazayas** [comment]
Hmm, the linter's still failing, I'll need to look at this one in a little more depth.

**Comment by jagthedrummer** [comment]
@gazayas I'm working on the ruby version problem. Will merge this as soon as I have the kinks worked out. https://github.com/bullet-train-co/bullet_train-core/pull/748

**Comment by gazayas** [comment]
@jagthedrummer I didn't tag you earlier when I merged the latest changes with `main`, just let me know if you need any more changes before merging!

**Comment by jagthedrummer** [comment]
@gazayas, I'm so sorry that I missed your update on this a couple of months ago. It looks like we have some more conflicts now. 😞

For some reason GitHub isn't letting me resolve them in the browser, so I just took a stab at rebasing all of this locally. I pushed it to a new branch and am letting CI run. https://github.com/bullet-train-co/bullet_train-core/pull/799/files

If that PR looks right (and if CI passes) I can push that branch onto this PR.

Sorry that I let this hang out so long. 🤦

**Comment by gazayas** [comment]
@jagthedrummer No problem! I just merged `main` again and the conflicts weren't so bad (mainly just comments I added here which weren't on `main`).

If you still want me to look at the rebase PR you submitted, I'd be glad to look at it.

**Comment by jagthedrummer** [comment]
Nah, don't worry about that other PR @gazayas. I'll go ahead and close that one now, and then this PR will be on the top of my list to get merged next week. Thanks again!

**Comment by jagthedrummer** [comment]
OK, @gazayas I just did some testing with this and we're getting really close! Here's what I did to test:

First I had my starter repo using the released versions of all the gems, and then I generated a `Project` model with a bunch of fields to test against.

```
bin/super-scaffold crud Project Team \
  name:text_field \
  logo:file_field \
  address:address_field \
  a_buttons:buttons \
  some_buttons:buttons{multiple} \
  a_color:color_picker \
  a_date_with_time:date_and_time_field \
  a_date:date_field an_option:options \
  some_options:options{multiple} \
  trix_editor_area:trix_editor \
  a_password:password_field \
  phone_number:phone_field \
  a_super_select:super_select \
  some_super_selects:super_select{multiple}
```

Then I ran `bin/hack --link ../bullet_train-core` to link my starter repo to this branch in my `core` repo.

And then I visited the form for editing a project, and I see two pairs (four total) deprecation messages in the log:

```
DEPRECATION WARNING: The `multiple` attribute will be removed in a later version.
  Please pass `options: {multiple: true}` to `render` instead.
  (called from render at /Users/jgreen/projects/bullet-train-co/bullet_train-core/bullet_train-themes/app/helpers/current_theme_helper.rb:19)
DEPRECATION WARNING: Calling warn on ActiveSupport::Deprecation is deprecated and will be removed from Rails
  (use your own Deprecation object instead)
  (called from ___sers_jgreen_projects_bullet_train_co_bullet_train_core_bullet_train_themes_tailwind_css_app_views_themes_tailwind_css_fields__buttons_html_erb__2608134009605983127_119560 at /Users/jgreen/projects/bullet-train-co/bullet_train-core/bullet_train-themes-tailwind_css/app/views/themes/tailwind_css/fields/_buttons.html.erb:38)


DEPRECATION WARNING: The `multiple` attribute will be removed in a later version.
  Please pass `options: {multiple: true}` to `render` instead.
  (called from render at /Users/jgreen/projects/bullet-train-co/bullet_train-core/bullet_train-themes/app/helpers/current_theme_helper.rb:19)
DEPRECATION WARNING: Calling warn on ActiveSupport::Deprecation is deprecated and will be removed from Rails
  (use your own Deprecation object instead)
  (called from ___sers_jgreen_projects_bullet_train_co_bullet_train_core_bullet_train_themes_tailwind_css_app_views_themes_tailwind_css_fields__options_html_erb___67058703469920398_119680 at /Users/jgreen/projects/bullet-train-co/bullet_train-core/bullet_train-themes-tailwind_css/app/views/themes/tailwind_css/fields/_options.html.erb:36)
```

As happened previously the warnings point to `theme_helper.rb` instead of to my project form due to `bullet_train-themes` being a local file. So I set my `Gemfile` to use the released version of `bullet_train-themes`.

After doing that the deprecation messages look like this:

```
DEPRECATION WARNING: The `multiple` attribute will be removed in a later version.
  Please pass `options: {multiple: true}` to `render` instead.
  (called from block (3 levels) in _app_views_account_projects__form_html_erb__4604713326196375126_119220 at /Users/jgreen/projects/bullet-train-co/bullet_train/app/views/account/projects/_form.html.erb:12)
DEPRECATION WARNING: Calling warn on ActiveSupport::Deprecation is deprecated and will be removed from Rails
  (use your own Deprecation object instead)
  (called from ___sers_jgreen_projects_bullet_train_co_bullet_train_core_bullet_train_themes_tailwind_css_app_views_themes_tailwind_css_fields__buttons_html_erb__956510538507361194_119580 at /Users/jgreen/projects/bullet-train-co/bullet_train-core/bullet_train-themes-tailwind_css/app/views/themes/tailwind_css/fields/_buttons.html.erb:38)



DEPRECATION WARNING: The `multiple` attribute will be removed in a later version.
  Please pass `options: {multiple: true}` to `render` instead.
  (called from block (3 levels) in _app_views_account_projects__form_html_erb__4604713326196375126_119220 at /Users/jgreen/projects/bullet-train-co/bullet_train/app/views/account/projects/_form.html.erb:17)
DEPRECATION WARNING: Calling warn on ActiveSupport::Deprecation is deprecated and will be removed from Rails
  (use your own Deprecation object instead)
  (called from ___sers_jgreen_projects_bullet_train_co_bullet_train_core_bullet_train_themes_tailwind_css_app_views_themes_tailwind_css_fields__options_html_erb___997636023711715181_119700 at /Users/jgreen/projects/bullet-train-co/bullet_train-core/bullet_train-themes-tailwind_css/app/views/themes/tailwind_css/fields/_options.html.erb:36)
```

So the deprecation warnings that we're raising are now being reported from the right place (`app/views/account/projects/_form.html.erb`), which is great! But there's an additional deprecation warning that _we_ are not raising, but is being raised by rails because of the way that we're raising our own deprecation warning.  (Yo dawg, we heard you like deprecation warnings... 😜)

Looks like instead of doing this:

```ruby
ActiveSupport::Deprecation.warn("our message")
```

We should do something more like this:

```ruby
ActiveSupport::Deprecation.new.warn("our message")
```

**Comment by jagthedrummer** [comment]
I just made the change mentioned above. Gonna let CI run, and then merge this sucker. Thanks for your work on this @gazayas! It's definitely going to make the whole `options` thing more consistent and therefore hopefully less confusing.

**Review by jagthedrummer — APPROVED** [review]

**Comment by gazayas** [comment]
> (Yo dawg, we heard you like deprecation warnings... 😜)
:laughing:

Thanks for picking up that last part on the deprecation warnings! I saw that there's a merge conflict (?) on #801 so I'll take a glance at that one, glad this one is merged though!

**Pull request merged**